### PR TITLE
Adds duplication check for source files in CI

### LINKED ISSUE #2076: Our CI should check that there are no duplicate filenames
If we have duplicated filenames, our include guard pattern (see #2075) could cause breakages.

### PR DESCRIPTION
**Pull request by innerout**
## Description

This PR checks for duplicate files in `src/catch2` to avoid include guard conflicts.
It adds a python script in the CI that recursively checks that all files are unique.  
## GitHub Issues
Closes #2076

### CODE REVIEW

File: check_duplicates.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+import os
+import sys
+
+files_map = {}
+
+for root, dir, files in os.walk("src/catch2"):
+    for file in files:
+        if file not in files_map:
+            files_map[file] = 1
```
Comment (horenmar) [inline check_duplicates.py]: Why not use a `set` and not have to make up a dummy key?

Alternatively, keep around the full path to the file, so the error can also report paths to the file... (But using plain set is simpler).

File: .github/workflows/check-duplicates.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
+name: Check duplicates
+
+# Run this workflow every time a new commit pushed to your repository
+
+on: [push, pull_request]
+
+jobs:
+  # Set the job key. The key is displayed as the job name
+  # when a job name is not provided
+  checkguard:
```
Comment (horenmar) [inline .github/workflows/check-duplicates.yml]: This should have a different key (`checknames`?), and be part of `validate-header-guards.yml`. Both of the jobs are part of validating the header guards, so let's keep them together. (It also removes some shared boilerplate.)

File: check_duplicates.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env python3
```
Comment (horenmar) [inline check_duplicates.py]: This file should be in `tools/scripts`, and probably have a more obvious filename (`checkDuplicateFilenames`?).

**Review by horenmar — CHANGES_REQUESTED** [review]

File: check_duplicates.py (hunk 1 above)
Comment (innerout) [reply check_duplicates.py]: I did not know python had sets thanks for telling me.

File: .github/workflows/check-duplicates.yml (hunk 2 above)
Comment (innerout) [reply .github/workflows/check-duplicates.yml]: Refactored `validate-header-guards.yml` to avoid duplicated code.

**Review by horenmar — APPROVED** [review]

**Comment by horenmar** [comment]
Looks good. Can you rebase + squash the changes?

**Comment by innerout** [comment]
Yes, I can rebase. Should I squash all the commits into the first?

**Comment by horenmar** [comment]
Yes

**Pull request merged**